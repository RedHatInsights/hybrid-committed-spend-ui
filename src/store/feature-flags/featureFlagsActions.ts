import { createAction } from 'typesafe-actions';

export interface FeatureFlagsActionMeta {
  isBillingStageFeatureEnabled?: boolean;
}

export const setFeatureFlags = createAction('feature/init_feature_flags')<FeatureFlagsActionMeta>();
export const resetState = createAction('feature/reset_state')();
