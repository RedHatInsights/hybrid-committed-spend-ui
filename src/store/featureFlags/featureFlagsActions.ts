import { createAction } from 'typesafe-actions';

export interface FeatureFlagsActionMeta {
  isExplorerFeatureEnabled?: boolean;
}

export const setFeatureFlags = createAction('feature/init_feature_flags')<FeatureFlagsActionMeta>();
export const resetState = createAction('feature/reset_state')();
