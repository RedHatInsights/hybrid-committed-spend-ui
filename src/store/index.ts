export { RootState } from './rootReducer';
export { configureStore } from './store';
