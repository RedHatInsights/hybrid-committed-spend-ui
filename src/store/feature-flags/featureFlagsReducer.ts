import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import type { resetState } from './featureFlagsActions';
import { setFeatureFlags } from './featureFlagsActions';

export type FeatureFlagsAction = ActionType<typeof setFeatureFlags | typeof resetState>;

export type FeatureFlagsState = Readonly<{
  hasFeatureFlags: boolean;
  isDetailsFeatureEnabled: boolean;
}>;

export const defaultState: FeatureFlagsState = {
  hasFeatureFlags: false,
  isDetailsFeatureEnabled: false,
};

export const stateKey = 'featureFlags';

export function featureFlagsReducer(state = defaultState, action: FeatureFlagsAction): FeatureFlagsState {
  switch (action.type) {
    case getType(setFeatureFlags):
      return {
        ...state,
        hasFeatureFlags: true,
        isDetailsFeatureEnabled: action.payload.isDetailsFeatureEnabled,
      };

    default:
      return state;
  }
}
