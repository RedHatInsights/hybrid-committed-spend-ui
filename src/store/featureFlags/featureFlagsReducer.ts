import { ActionType, getType } from 'typesafe-actions';

import { resetState, setFeatureFlags } from './featureFlagsActions';

export type FeatureFlagsAction = ActionType<typeof setFeatureFlags | typeof resetState>;

export type FeatureFlagsState = Readonly<{
  isExplorerFeatureEnabled: boolean;
}>;

export const defaultState: FeatureFlagsState = {
  isExplorerFeatureEnabled: false,
};

export const stateKey = 'featureFlags';

export function featureFlagsReducer(state = defaultState, action: FeatureFlagsAction): FeatureFlagsState {
  switch (action.type) {
    case getType(resetState):
      state = defaultState;
      return state;

    case getType(setFeatureFlags):
      return {
        ...state,
        isExplorerFeatureEnabled: action.payload.isExplorerFeatureEnabled,
      };

    default:
      return state;
  }
}
