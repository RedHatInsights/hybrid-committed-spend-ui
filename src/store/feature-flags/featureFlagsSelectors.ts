import type { RootState } from 'store/rootReducer';

import { stateKey } from './featureFlagsReducer';

export const selectFeatureFlagsState = (state: RootState) => state[stateKey];

export const selectHasFeatureFlags = (state: RootState) => selectFeatureFlagsState(state).hasFeatureFlags;

export const selectIsBillingStageFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isBillingStageFeatureEnabled;
