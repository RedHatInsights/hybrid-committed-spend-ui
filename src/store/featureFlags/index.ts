import * as featureFlagsActions from './featureFlagsActions';
import {
  FeatureFlagsAction,
  featureFlagsReducer,
  FeatureFlagsState,
  stateKey as featureFlagsStateKey,
} from './featureFlagsReducer';
import * as featureFlagsSelectors from './featureFlagsSelectors';

export {
  FeatureFlagsAction,
  featureFlagsActions,
  featureFlagsReducer,
  featureFlagsSelectors,
  FeatureFlagsState,
  featureFlagsStateKey,
};
