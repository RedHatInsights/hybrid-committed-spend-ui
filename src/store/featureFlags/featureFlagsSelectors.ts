import { RootState } from 'store/rootReducer';

import { stateKey } from './featureFlagsReducer';

export const selectFeatureFlagsState = (state: RootState) => state[stateKey];

export const selectIsExplorerFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isExplorerFeatureEnabled;
