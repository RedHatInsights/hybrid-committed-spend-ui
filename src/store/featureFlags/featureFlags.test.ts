import { createMockStoreCreator } from 'store/mockStore';

import { featureFlagsSelectors } from '.';
import * as actions from './featureFlagsActions';
import { featureFlagsReducer, stateKey } from './featureFlagsReducer';
import * as selectors from './featureFlagsSelectors';

const createUIStore = createMockStoreCreator({
  [stateKey]: featureFlagsReducer,
});

test('default state', async () => {
  const store = createUIStore();
  expect(selectors.selectFeatureFlagsState(store.getState())).toMatchSnapshot();
});

test('sample feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isExplorerFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsExplorerFeatureEnabled(store.getState())).toBe(true);
});
