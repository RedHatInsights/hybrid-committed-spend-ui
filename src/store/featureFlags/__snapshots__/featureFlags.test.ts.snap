// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
{
  "isCurrencyFeatureEnabled": false,
  "isExportsFeatureEnabled": false,
  "isIbmFeatureEnabled": false,
  "isOciFeatureEnabled": false,
}
`;
